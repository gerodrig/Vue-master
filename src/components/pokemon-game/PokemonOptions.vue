<template>
    <div class="bg-black opacity-80">
        <div class="flex items-center mx-auto justify-center max-w-[800px]">
            <ul class="flex-wrap justify-center py-6 text-blue-700 list-none md:flex">
            <li
                v-for="{ name, id } in options"
                @click="$emit('chosenPokemon', id)"
                :class="{ 'opacity-50': disabled }"
                :style="{ 'pointer-events': disabled ? 'none' : 'auto' }"
                :aria-disabled="disabled"
                :key="id">{{
                name.toLocaleUpperCase()
                }}</li>
            </ul>
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent, toRef } from 'vue';
import { Pokemon } from '@interfaces/pokemon-game/index';

export default defineComponent({
  name: 'PokemonOptions',
  props: {
    options: {
      type: Array as () => Pokemon[],
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    
    const options = toRef(props, 'options');
    const disabled = toRef(props, 'disabled');

    //upercase first letter
    options.value.forEach((option) => {
      option.name = option.name[0].toUpperCase() + option.name.slice(1);
    });
    return {
      options,
      disabled,
    };
  },
});
</script>

<style scoped>
li {
  @apply bg-white rounded-md border border-gray-300 cursor-pointer mb-3 w-96 hover:bg-gray-100 text-center py-2 md:mr-3;
}
</style>
