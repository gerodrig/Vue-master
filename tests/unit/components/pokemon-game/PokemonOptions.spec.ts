import {describe, it, expect, beforeEach } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import PokemonOptions from '../../../../src/components/pokemon-game/PokemonOptions.vue';


import { pokemons } from '../../mocks/pokemon-game/pokemons.mock';

describe('PokemonOptions Component', () => {

    let wrapper: any;

    beforeEach(() => {
        wrapper = shallowMount(PokemonOptions, {
            props: {
                options: pokemons
            }
        });
    });

  it('must match the snapshot', () => {

    // console.log(wrapper.html());

    expect(wrapper.html()).toMatchSnapshot();
    
  });

  it('should display 4 options correctly', () => {
    //make sure that 4 list items are rendered
    //each li must have the correct pokemon name

    const liElements = wrapper.findAll('li');

    //console.log(liElements[0].text());
    expect(liElements).toHaveLength(4);
    expect(liElements[0].text()).toBe('bulbasaur'.toLocaleUpperCase());
    expect(liElements[1].text()).toBe('ivysaur'.toLocaleUpperCase());
    expect(liElements[2].text()).toBe('venusaur'.toLocaleUpperCase());
    expect(liElements[3].text()).toBe('charmander'.toLocaleUpperCase());

  });

  it('should emit "onSelectPokemon" with the id of the pokemon', () => {
    //make sure that the event is emitted with the correct payload
    const [li1, li2, li3, li4] = wrapper.findAll('li');

    li1.trigger('click');
    li2.trigger('click');
    li3.trigger('click');
    li4.trigger('click');

    const event = wrapper.emitted('chosenPokemon');

    // console.log(event);

    // expect(event).toBe(1);
    expect(event.length).toBe(4);
    expect(event[0]).toEqual([1]);
    expect(event[1]).toEqual([2]);
    expect(event[2]).toEqual([3]);
    expect(event[3]).toEqual([4]);
  });
});