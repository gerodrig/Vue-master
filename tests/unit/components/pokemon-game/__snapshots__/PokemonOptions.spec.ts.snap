// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonOptions Component > must match the snapshot 1`] = `
"<div data-v-6c7a75d2=\\"\\" class=\\"bg-black opacity-80\\">
  <div data-v-6c7a75d2=\\"\\" class=\\"flex items-center mx-auto justify-center max-w-[800px]\\">
    <ul data-v-6c7a75d2=\\"\\" class=\\"flex-wrap justify-center py-6 text-blue-700 list-none md:flex\\">
      <li data-v-6c7a75d2=\\"\\" class=\\"\\" style=\\"pointer-events: auto;\\" aria-disabled=\\"false\\">BULBASAUR</li>
      <li data-v-6c7a75d2=\\"\\" class=\\"\\" style=\\"pointer-events: auto;\\" aria-disabled=\\"false\\">IVYSAUR</li>
      <li data-v-6c7a75d2=\\"\\" class=\\"\\" style=\\"pointer-events: auto;\\" aria-disabled=\\"false\\">VENUSAUR</li>
      <li data-v-6c7a75d2=\\"\\" class=\\"\\" style=\\"pointer-events: auto;\\" aria-disabled=\\"false\\">CHARMANDER</li>
    </ul>
  </div>
</div>"
`;
